src/compas_ags/ags/graphstatics.py

```diff
@@ -337,6 +351,82 @@ def form_update_from_force(form, force, kmax=100):
         attr['l'] = forces[index, 0]
 
 
+def form_update_from_force_newton(form, force, constraints=None, tol=1e-10, max_iter=20):
+    r"""Update the form diagram after a modification of the force diagram.
+
+    Compute the geometry of the form diagram from the geometry of the force diagram
+    and some constraints, including non-linear.
+    The form diagram is computed by formulating the root-finding approach presented
+    in bi-dir AGS [1]. Newton's method is used to solve for the form diagram.
+
+    The algorithm fails if it is over-constrained or if the initial guess is too far
+    from any root.
+
+    Parameters
+    ----------
+    form : FormDiagram
+        The form diagram to update.
+    force : ForceDiagram
+        The force diagram containing the vertex modification desired.
+    constraints : ConstraintsCollection (None)
+        A collection of form diagram constraints.
+    tol: float (1e-10)
+        Stopping criteria tolerance.
+    max_iter : int (20)
+        Maximum number of iterations before stop Newton Method.
+
+    Returns
+    -------
+    None
+        The form and force diagram are updated in-place.
+
+    Reference
+    ----------
+        [1] Alic, V. and Åkesson, D., 2017. Bi-directional algebraic graphic statics. Computer-Aided Design, 93, pp.26-37.
+
+    Examples
+    --------
+    >>>
+
+    """
+    X = array(form.vertices_attribute('x') + form.vertices_attribute('y')).reshape(-1, 1)
+    _X_goal = array(force.vertices_attribute('x') + force.vertices_attribute('y')).reshape(-1, 1)
+
+    vcount = form.number_of_vertices()
+    index_vertex = form.index_vertex()
+
+    # Begin Newton
+    diff = 100
+    n_iter = 1
+    while diff > tol:
+        # Update force diagram based on form at each iteration
+        form_update_q_from_qind(form)
+        force_update_from_form(force, form)
+
+        # Get jacobian maxtrix and residual vector considering constraints
+        red_jacobian, red_r = get_red_residual_and_jacobian(form, force, _X_goal, constraints)
+
+        # Do the least squares solution
+        dx = lstsq(red_jacobian, -red_r)[0]
+
+        X = X + dx
+
+        # Update form diagram at end of iteration
+        for i in range(vcount):
+            vertex = index_vertex[i]
+            form.vertex_attribute(vertex, 'x', X[i].item())
+            form.vertex_attribute(vertex, 'y', X[i + vcount].item())
```

Comment:
If I don't do `.item` this breaks because `form.xy()`  returns something nonsensical, with a weird dimension, instead of (n, 2) the dimension becomes (n,2,1,1).
`form.xy()` is used afterwards in `form_from_qind`

Reply:
it is necessary